src/java/org/apache/cassandra/io/sstable/format/bti/RowIndexWriter.java

```
@@ -0,0 +1,122 @@
+    public void close()
+    {
+        trie.close();
+    }
+
+    void add(ClusteringPrefix<?> firstName, ClusteringPrefix<?> lastName, IndexInfo info) throws IOException
+    {
+        assert info.openDeletion != null;
+        ByteComparable sep = prevMax == null
+                             ? ByteComparable.EMPTY
+                             : ByteComparable.separatorGt(prevMax, comparator.asByteComparable(firstName));
+        trie.add(sep, info);
+        prevSep = sep;
+        prevMax = comparator.asByteComparable(lastName);
+    }
+
+    public long complete(long endPos) throws IOException
+    {
+        // Add a separator after the last section, so that greater inputs can be quickly rejected.
+        // To maximize its efficiency we add it with the length of the last added separator.
+        int i = 0;
+        ByteSource max = prevMax.asComparableBytes(Walker.BYTE_COMPARABLE_VERSION);
+        ByteSource sep = prevSep.asComparableBytes(Walker.BYTE_COMPARABLE_VERSION);
+        int c;
+        while ((c = max.next()) == sep.next() && c != ByteSource.END_OF_STREAM)
+            ++i;
+        assert c != ByteSource.END_OF_STREAM : "Corrupted row order, max=" + prevMax;
+
+        trie.add(nudge(prevMax, i), new IndexInfo(endPos, DeletionTime.LIVE));
+
+        return trie.complete();
+    }
+
+    /**
+     * Produces a source that is slightly greater than argument with length at least nudgeAt.
+     */
+    private ByteComparable nudge(ByteComparable value, int nudgeAt)
+    {
+        return version -> new ByteSource()
+        {
+            private final ByteSource v = value.asComparableBytes(version);
+            private int cur = 0;
+
+            @Override
+            public int next()
+            {
+                int b = ByteSource.END_OF_STREAM;
+                if (cur <= nudgeAt)
+                {
+                    b = v.next();
+                    if (cur == nudgeAt)
+                    {
+                        if (b < 255)
```

nit: 255 is canonically the max value for an unsigned byte I guess, but making it a constant somewhere wouldn't be terrible. Not super important...

---

I personally prefer not to hide fundamental constants behind identifiers as they can mislead and hide errors.

Changed to `0xFF` to make it even more obvious.

---

Works for me.